posthog/async_migrations/migrations/0001_events_sample_by.py

```diff
@@ -0,0 +1,163 @@
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+            INSERT INTO {TEMPORARY_TABLE_NAME}
+            SELECT * 
+            FROM {EVENTS_TABLE}
+            WHERE timestamp < toDateTime64(toYYYYMM(now()) - 1, 6)
+            AND timestamp >= (SELECT max(timestamp) FROM {TEMPORARY_TABLE_NAME})""",
+            rollback=f"TRUNCATE TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"DETACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=f"ATTACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            side_effect=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", False),
+            side_effect_rollback=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", True),
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+            INSERT INTO {TEMPORARY_TABLE_NAME}
+            SELECT * 
+            FROM {EVENTS_TABLE}
+            WHERE timestamp >= (SELECT max(timestamp) FROM {TEMPORARY_TABLE_NAME})""",
+            rollback=f"TRUNCATE TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"""
+                RENAME TABLE
+                    {EVENTS_TABLE_NAME} to {BACKUP_TABLE_NAME},
+                    {TEMPORARY_TABLE_NAME} to {EVENTS_TABLE_NAME}
+                ON CLUSTER {CLICKHOUSE_CLUSTER}
+            """,
+            rollback=f"""
+                RENAME TABLE
+                    {EVENTS_TABLE_NAME} to {EVENTS_TABLE_NAME}_failed,
+                    {BACKUP_TABLE_NAME} to {EVENTS_TABLE_NAME}
+                ON CLUSTER {CLICKHOUSE_CLUSTER}
+            """,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"OPTIMIZE TABLE {EVENTS_TABLE_NAME} FINAL",
+            rollback="",
+            resumable=True,
+        ),
+        AsyncMigrationOperation(
+            database=AnalyticsDBMS.CLICKHOUSE,
+            sql=f"ATTACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            rollback=f"DETACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
+            side_effect=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", True),
+            side_effect_rollback=lambda: setattr(config, "MATERIALIZED_COLUMNS_ENABLED", False),
+        ),
+    ]
+
+    def is_required(self):
+        res = sync_execute(f"SHOW CREATE TABLE {EVENTS_TABLE_NAME}")
+        return "ORDER BY (team_id, toDate(timestamp), cityHash64(distinct_id), cityHash64(uuid))" not in res[0][0]
+
+    def precheck(self):
```

Comment:
If possible it would be great if we could check Kafka retention as well & make sure we have a couple of days, in case things go wrong and ingestion is down longer to make it less probable for self-hosted users to lose data